# IMPORTING DEPENDENCIES

In [15]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn import svm
from sklearn.metrics import accuracy_score
import matplotlib.pyplot as plt
import seaborn as sns


In [16]:
pip install kagglehub[pandas-datasets]

In [17]:
!pip install kagglehub
from google.colab import files
files.upload()  # Upload your kaggle.json here
!mkdir ~/.kaggle
!cp kaggle.json ~/.kaggle/
!chmod 600 ~/.kaggle/kaggle.json

Saving kaggle.json to kaggle (1).json
mkdir: cannot create directory ‘/root/.kaggle’: File exists


In [18]:
import kagglehub

# Download latest version
path = kagglehub.dataset_download("mlg-ulb/creditcardfraud")

print("Path to dataset files:", path)


Path to dataset files: /root/.cache/kagglehub/datasets/mlg-ulb/creditcardfraud/versions/3


In [19]:
from pathlib import Path

# Construct the full path to the CSV file
csv_path = Path(path) / "creditcard.csv"

# Load the CSV file into a pandas DataFrame
credit_card_data = pd.read_csv(csv_path)

# Display the first 5 rows of the dataset
print("First 5 records:")
print(credit_card_data.head())

First 5 records:
   Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26    

In [20]:
credit_card_data.tail()

Time         V1         V2        V3        V4        V5  \
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[5 rows x 31 columns]

In [21]:
# Get Information about Dataset
credit_card_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [22]:
# Checking the Number of missing values in each column
credit_card_data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [23]:
# distribution of legit Transaction & Fraudulent Transactions
credit_card_data['Class'].value_counts()


Class
0    284315
1       492
Name: count, dtype: int64

Above Dataset is highly imbalanced, with only 492 fraud cases (Class 1) out of 284,807 total transactions.

This imbalance can cause issues when training machine learning models, as they may become biased toward the majority class (non-fraud). Here's how you can preprocess and handle this imbalance:

## 1. Understand the Imbalance

Class Distribution:

1. Non-fraud (Class 0): 284,315

2. Fraud (Class 1): 492

Imbalance Ratio: ~578:1 (very severe imbalance)

### **So we need to Handle the Imbalanced data**

Before addressing the imbalance, split the dataset into training and testing sets to avoid data leakage:

In [24]:
# seperating the data for analysis

# Separate features (X) and target (y)
legit = credit_card_data.drop("Class", axis=1)
fraud = credit_card_data["Class"]

# Split into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(legit, fraud, test_size=0.2, random_state=42, stratify=fraud)
#Why stratify= fraud? Ensures the same class distribution is maintained in both training and testing sets.

print("Training set shape:", X_train.shape)
print("Testing set shape:", X_test.shape)

Training set shape: (227845, 30)
Testing set shape: (56962, 30)


In [25]:
# statical measures of the data
legit.Amount.describe()

count    284807.000000
mean         88.349619
std         250.120109
min           0.000000
25%           5.600000
50%          22.000000
75%          77.165000
max       25691.160000
Name: Amount, dtype: float64

 ## Handle Imbalanced Data

There are several techniques to handle imbalanced datasets.

1.   Oversampling the Minority Class: Use techniques like SMOTE (Synthetic Minority Oversampling Technique) to generate synthetic samples for the minority class.


2.   Undersampling the Majority Class: Randomly remove samples from the majority class to balance the dataset.




In [26]:
from imblearn.under_sampling import RandomUnderSampler

undersampler = RandomUnderSampler(random_state=42)
X_train_res, y_train_res = undersampler.fit_resample(X_train, y_train)

print("Resampled training set shape:", X_train_res.shape)
print("Resampled class distribution:\n", y_train_res.value_counts())

Resampled training set shape: (788, 30)
Resampled class distribution:
 Class
0    394
1    394
Name: count, dtype: int64


Concatanating 2 Dataframes

In [27]:
new_dataset = pd.concat([X_train_res, y_train_res], axis=1)
new_dataset.head()

Time        V1        V2        V3        V4        V5        V6  \
20945    31369.0  1.114107  0.188459  0.149144  1.153868  0.224011  0.396246   
159516  112588.0  1.910941  0.686774 -0.595770  3.817385  0.288991 -0.660187   
229456  145930.0  0.013514  0.452287 -0.738086 -0.950309  1.405159 -0.423082   
127881   78505.0 -1.273388  0.270815  1.256719  1.064659 -1.401545  0.234430   
158110  110944.0  1.978615  0.114560 -1.322450  0.577586  0.348417 -0.376863   

              V7        V8        V9  ...       V21       V22       V23  \
20945   0.020168  0.217603 -0.141379  ...  0.065655  0.367010 -0.090954   
159516  0.365387 -0.261102 -1.226056  ...  0.266636  0.805442  0.114189   
229456  1.115085 -0.073982 -0.349808  ...  0.152423  0.411430 -0.017938   
127881  1.194255  0.412955 -0.528987  ...  0.407454  0.673393  0.759378   
158110 -0.211046 -0.123861  1.623757  ... -0.138925 -0.031185  0.191191   

             V24       V25       V26       V27       V28  Amount  Class  
20945  -0.271464  0.646597 -0.217493  0.035881 -0.004553    1.00      0  
159516  0.401013  0.069967  0.108304 -0.039724 -0.047338    5.14      0  
229456  0.227643 -0.857711  0.444818  0.136231  0.229043   22.95      0  
127881  0.580046 -0.310258 -0.463227 -0.023899  0.134970  325.00      0  
158110  0.598530 -0.182579  0.389251 -0.075180 -0.040764   14.05      0  

[5 rows x 31 columns]

In [29]:
new_dataset['Class'].value_counts()

Class
0    394
1    394
Name: count, dtype: int64

In [30]:
new_dataset.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94068.428934  0.086256 -0.063777  0.069997  0.033105 -0.076524   
1      82832.177665 -4.707808  3.588729 -7.068378  4.592975 -3.101629   

             V6        V7       V8        V9  ...       V20       V21  \
Class                                         ...                       
0     -0.014895  0.050129  0.00886  0.003166  ...  0.048927  0.015806   
1     -1.387192 -5.539909  0.58792 -2.589654  ...  0.358018  0.628814   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0      0.053978  0.006621  0.008088 -0.024533 -0.012803  0.001019  0.005723   
1      0.051318 -0.062790 -0.109108  0.019602  0.047827  0.155933  0.077212   

           Amount  
Class              
0      105.160482  
1      125.591472  

[2 rows x 30 columns]

## Split the Concatenated Dataset

Since new_dataset contains both features and the target, we need to split it back into X (features) and y (target) for training:



In [32]:
# Separate features (X) and target (y)
X_train_balanced = new_dataset.drop("Class", axis=1)  # Features
y_train_balanced = new_dataset["Class"]  # Target

print("Balanced training set shape:", X_train_balanced)
print("Balanced class distribution:\n", y_train_balanced)

Balanced training set shape:             Time         V1         V2         V3        V4         V5  \
20945    31369.0   1.114107   0.188459   0.149144  1.153868   0.224011   
159516  112588.0   1.910941   0.686774  -0.595770  3.817385   0.288991   
229456  145930.0   0.013514   0.452287  -0.738086 -0.950309   1.405159   
127881   78505.0  -1.273388   0.270815   1.256719  1.064659  -1.401545   
158110  110944.0   1.978615   0.114560  -1.322450  0.577586   0.348417   
...          ...        ...        ...        ...       ...        ...   
223366  143354.0   1.118331   2.074439  -3.837518  5.448060   0.071816   
20198    30852.0  -2.830984   0.885657   1.199930  2.861292   0.321669   
261473  160034.0  -2.349340   1.512604  -2.647497  1.753792   0.406328   
151011   94364.0 -15.192064  10.432528 -19.629515  8.046075 -12.838167   
120505   75851.0  -4.793667   3.418911  -5.074445  4.035987  -3.527875   

              V6         V7        V8        V9  ...       V20       V21  \
20945 

## Train a Model

Now, train a model on the balanced dataset. Let's use a Random Forest Classifier:

In [33]:
from sklearn.ensemble import RandomForestClassifier

# Initialize the model
model = RandomForestClassifier(random_state=42)

# Train the model on the balanced data
model.fit(X_train_balanced, y_train_balanced)

RandomForestClassifier(random_state=42)

## Evaluate the Model on the Test Set


Evaluate the model on the original (unresampled) test set (X_test, y_test) to ensure the model's performance is tested on realistic, imbalanced data.

In [34]:
# Predict on the test set
y_pred = model.predict(X_test)

 ### Evaluate Using Metrics for Imbalanced Data

Use metrics like Confusion Matrix, Precision, Recall, F1-Score, and ROC-AUC to evaluate the model:

In [35]:
from sklearn.metrics import confusion_matrix, classification_report, roc_auc_score

# Confusion Matrix
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))

# Classification Report (Precision, Recall, F1-Score)
print("Classification Report:\n", classification_report(y_test, y_pred))

# ROC-AUC Score
print("ROC-AUC Score:", roc_auc_score(y_test, y_pred))

Confusion Matrix:
 [[54827  2037]
 [    8    90]]
Classification Report:
               precision    recall  f1-score   support

           0       1.00      0.96      0.98     56864
           1       0.04      0.92      0.08        98

    accuracy                           0.96     56962
   macro avg       0.52      0.94      0.53     56962
weighted avg       1.00      0.96      0.98     56962

ROC-AUC Score: 0.9412725170259437


### Accuracy score

In [40]:
X_train_prediction = model.predict(X_train_balanced)
training_data_accuracy = accuracy_score(X_train_prediction, y_train_balanced)

In [41]:
print('Accuracy on Training data: ', training_data_accuracy)

Accuracy on Training data:  1.0


In [43]:
# Accuracy on test data
X_test_prediction = model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, y_test)

In [48]:
print('Accuracy on test data: ', accuracy_score(y_test, y_pred))

Accuracy on test data:  0.9640988729328324


## Improve the Model (Optional)

If the model's performance is not satisfactory, we can try the following:

1.  Hyperparameter Tuning
Use GridSearchCV or RandomizedSearchCV to find the best hyperparameters for your model:

In [37]:
from sklearn.model_selection import GridSearchCV

# Define parameter grid
param_grid = {
    'n_estimators': [100, 200],
    'max_depth': [None, 10, 20],
    'min_samples_split': [2, 5]
}

# Perform grid search
grid_search = GridSearchCV(RandomForestClassifier(random_state=42), param_grid, cv=3, scoring='roc_auc')
grid_search.fit(X_train_balanced, y_train_balanced)

# Best parameters
print("Best Parameters:", grid_search.best_params_)

# Train with best parameters
best_model = grid_search.best_estimator_

Best Parameters: {'max_depth': None, 'min_samples_split': 5, 'n_estimators': 100}


In [38]:
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))

# Classification Report (Precision, Recall, F1-Score)
print("Classification Report:\n", classification_report(y_test, y_pred))

# ROC-AUC Score
print("ROC-AUC Score:", roc_auc_score(y_test, y_pred))

Confusion Matrix:
 [[54827  2037]
 [    8    90]]
Classification Report:
               precision    recall  f1-score   support

           0       1.00      0.96      0.98     56864
           1       0.04      0.92      0.08        98

    accuracy                           0.96     56962
   macro avg       0.52      0.94      0.53     56962
weighted avg       1.00      0.96      0.98     56962

ROC-AUC Score: 0.9412725170259437


 ## Save the Model

Once we're satisfied with the model, save it for future use

In [39]:
import joblib

# Save the model
joblib.dump(model, "fraud_detection_model.pkl")

# Load the model (later)
loaded_model = joblib.load("fraud_detection_model.pkl")